allennlp/fairness/fairness_metrics.py

```diff
@@ -0,0 +1,940 @@
+            of the pairwise gaps (where the gap corresponding to prediction label i is at index i).
+            A gap of nearly 0 implies fairness on the basis of Demographic Parity in the Absence of Ground Truth.
+
+        !!! Note
+            If a possible class label is not present in Y, the expected behavior is that
+            the gaps corresponding to this class label are NaN. If a possible (class label,
+            protected variable label) pair is not present in the joint of Y and X, the expected
+            behavior is that the gap corresponding to this (class label, protected variable label)
+            pair is NaN.
+        """
+        gaps = {}
+        for x in range(self._num_protected_variable_labels):
+            gaps[x] = self.gap_func(x)
+        if reset:
+            self.reset()
+        return gaps
+
+    @overrides
+    def reset(self) -> None:
+        self._joint_counts_by_protected_variable_label = {
+            x: torch.zeros(self._num_classes) for x in range(self._num_protected_variable_labels)
+        }
+        self._protected_variable_label_counts = {
+            x: torch.tensor(0) for x in range(self._num_protected_variable_labels)
+        }
+        self._y_counts = torch.zeros(self._num_classes)
+        self._total_predictions = torch.tensor(0)
+
+    def _ova_gap(self, x: int):
+        device = self._y_counts.device
+        pmi_terms = self._all_pmi_terms()
+        pmi_not_x = 0.0
+        for not_x in range(self._num_protected_variable_labels):
+            if not_x == x:
+                continue
+            pmi_not_x += pmi_terms[not_x]
+        pmi_not_x /= self._num_protected_variable_labels - 1
+
+        # Will contain NaN if not all possible class labels are predicted
+        # Will contain NaN if not all possible (class label,
+        # protected variable label) pairs are predicted
+        gap = pmi_terms[x] - pmi_not_x
+        return torch.where(~gap.isinf(), gap, torch.tensor(float("nan")).to(device))
+
+    def _pairwise_gaps(self, x: int):
+        device = self._y_counts.device
+        pmi_terms = self._all_pmi_terms()
+        pairwise_gaps = {}
+        for not_x in range(self._num_protected_variable_labels):
+            gap = pmi_terms[x] - pmi_terms[not_x]
+            pairwise_gaps[not_x] = torch.where(
+                ~gap.isinf(), gap, torch.tensor(float("nan")).to(device)
+            )
+        return pairwise_gaps
+
+    def _all_pmi_terms(self) -> Dict[int, torch.Tensor]:
+        if self._total_predictions == 0:
+            return {
+                x: torch.full((self._num_classes,), float("nan"))
+                for x in range(self._num_protected_variable_labels)
+            }
+
+        device = self._y_counts.device
+        pmi_terms = {}
+        prob_y = torch.zeros(self._num_classes).to(device)
+        torch.div(self._y_counts, self._total_predictions, out=prob_y)
+        for x in range(self._num_protected_variable_labels):
```

Comment:
I think it's possible to vectorize this part if `joint_counts_by_protected_variable_label` and `_protected_variable_label_counts` are initialized as tensors instead of dicts (they can still be filled using the for loop like now).

Reply:
Great suggestion :) I just did this for all the metrics.